include/pybind11/detail/internals.h

```diff
@@ -328,6 +328,19 @@ const char *c_str(Args &&...args) {
     return strings.front().c_str();
 }
 
+// Thread state manipulation C API should not be called while Python runtime is finalizing.
+// For more detail see https://docs.python.org/3.7/c-api/init.html#c.PyEval_RestoreThread
+// `finalization_guard()` provides a version agnostic way to check if runtime is finalizing.
+inline bool finalization_guard() {
```

Comment:
This should definitely not be in `internals.h`, though. This is not related to the pybind11 internals (which have a specific meaning, being linked to the internal information pybind11 stores about types and instances).

Please move it back to somewhere related to the threads. Or potentially `common.h`, if this could be used in other places; but I don't think so, since the comment you wrote above is very specific to the threading functionality.

I would also want `finalization_guard` to get a better name, showing that it's linked to these threads, such that it doesn't get reused for other things, since it doesn't report "finalizing" correctly on Python 2? Or is there no concept of "finalizing" on Python 2?

Reply:
@YannickJadoul , how about `internal::is_finalizing()`? And I don't know much about Python-2 runtime, but quick grep thru the headers suggests that there isn't any.

Reply:
No, we have no `internal` namespace, so `detail`, you mean?

Then, yes, perhaps, but then the documentation you have on top should partially move back to where you use it? "Thread state manipulation C API should not be called while Python runtime is finalizing." (and accompanying link, etc.) does not belong to `is_finalizing`, then, but to where you use it, I assume?

Reply:
Also, now that I think about it: if we don't know how to detect whether Python 2 is finalizing, this is a reasonably dangerous API to create, no?

Reply:
Sorry, that was a typo, I've meant `detail::is_finalizing()`. 
As for Python-2: as there are no concept of runtime finalization in Python 2, therefore `is_finalizing()` should always return false to indicate that all C APIs are safe to call all the time.

Reply:
Here's `Py_Finalize` in 2.7: https://github.com/python/cpython/blob/2.7/Python/pythonrun.c#L411

I indeed can't immediately spot some finalizing flag being set.

Reply:
(While I can, on `master`: https://github.com/python/cpython/blob/296a796951032f678d063008f588ccc6958d0df1/Python/pylifecycle.c#L1636)

Reply:
> Here's `Py_Finalize` in 2.7: https://github.com/python/cpython/blob/2.7/Python/pythonrun.c#L411
> I indeed can't immediately spot some finalizing flag being set.

Sorry, I've meant there are no finalization flag/API that Python-2 exposes and there are doesn't seem to be any checks in `PyGILState_Ensure()` that prohibit calling this function while/after `Py_Finalize()` has been called.

Reply:
Perfect! Thanks for making sure :-)

Reply:
(That's what I also meant, btw, that I couldn't find any evidence of Python 2 having a "finalizing" state)